distributed/protocol/pickle.py

```diff
@@ -23,36 +24,46 @@ def _always_use_pickle_for(x):
         return False
 
 
-def dumps(x):
+def dumps(x, *, buffer_callback=None):
     """ Manage between cloudpickle and pickle
 
     1.  Try pickle
     2.  If it is short then check if it contains __main__
     3.  If it is long, then first check type, then check __main__
     """
+    buffers = []
+    dump_kwargs = {"protocol": HIGHEST_PROTOCOL}
+    if HIGHEST_PROTOCOL >= 5 and buffer_callback is not None:
+        dump_kwargs["buffer_callback"] = buffers.append
     try:
-        result = pickle.dumps(x, protocol=pickle.HIGHEST_PROTOCOL)
+        del buffers[:]
+        result = pickle.dumps(x, **dump_kwargs)
         if len(result) < 1000:
             if b"__main__" in result:
-                return cloudpickle.dumps(x, protocol=pickle.HIGHEST_PROTOCOL)
-            else:
-                return result
+                del buffers[:]
+                result = cloudpickle.dumps(x, **dump_kwargs)
         else:
-            if _always_use_pickle_for(x) or b"__main__" not in result:
-                return result
-            else:
-                return cloudpickle.dumps(x, protocol=pickle.HIGHEST_PROTOCOL)
+            if not (_always_use_pickle_for(x) or b"__main__" not in result):
+                del buffers[:]
+                result = cloudpickle.dumps(x, **dump_kwargs)
     except Exception:
         try:
-            return cloudpickle.dumps(x, protocol=pickle.HIGHEST_PROTOCOL)
+            del buffers[:]
+            result = cloudpickle.dumps(x, **dump_kwargs)
         except Exception as e:
             logger.info("Failed to serialize %s. Exception: %s", x, e)
             raise
+    for b in buffers:
+        buffer_callback(b)
```

Comment:
`buffer_callback` could be `None` per the signature, it's just never not none in usage.

```suggestion
    if buffer_callback is not None:
        for b in buffers:
            buffer_callback(b)
```

Reply:
Just noticed that `buffers` will never be appended to if `buffer_callback` is None, so the original code is correct. I find the explicit `is not None` check more readable, but up to you.

Reply:
Yeah tried both ways and went with the current form as it was less verbose. Happy to change if you prefer.

Reply:
In this case I think an explicit check is easier to understand. It took me a sec to fully understand the current logic.

Reply:
Yeah that makes sense. Included the change. Thanks!